File: src/conductances/K_Pst.py
"""
K_Pst: Slow potassium conductance

Extracted from BBP NEURON mechanism: K_Pst.mod
Neuron type: pyramidal neuron

State variables:
    m: activation gate
    h: inactivation gate
"""

# temperature correction
QT = 2.3**((34 - 21) / 10)

# euler's constant to 30 digits of precision
e = 2.718281828459045235360287471352

# maximal peak conductance density of channel per unit membrane area (S/cm^2)
gK_Pstbar = 0.00001

def derivatives(state, V):
    """
    Compute steady-state values and time constants,
    then compute time derivatives of the gating variables.

    Inputs:
        - state (tuple of floats): (m, h) gating variables
        - V (float): membrane potential (mV)

    Returns: 
        - dm_dt (float): time derivative of m
        - dh_dt (float): float time derivative of h
    """
    # voltage shift
    v = V + 10.0

    # steady state activation & inactivation
    mInf = 1.0 / (1.0 + e**(-(v + 1.0) / 12.0))
    hInf = 1.0 / (1.0 + e**(-(v + 54.0) / -11.0))

    # activation & inactivation time constants (ms)
    if v < -50:
        mTau = (1.25 + 175.03 * e**(0.026 * v)) / QT
    else:
        mTau = (1.25 + 13 * e**(-v * 0.026)) / QT

    hTau = (360 + (1010 + 24 * (v+55)) * e**(-((v + 75) / 48) ^ 2)) / QT

    m, h = state

    dm_dt = (mInf - m) / mTau
    dh_dt = (hInf - h) / hTau

    return dm_dt, dh_dt

def current(state, V, E_K, dt=0.025):
    """
    Compute current based on gating variables and membrane potential.

    Inputs:
        - state (tuple of floats): (m, h) gating variables
        - V (float): membrane potential (mV)
        - E_K (float): potassium reversal potential (mV)
        - time step (float): time step of derivative measurement

    Returns: 
        - I_K (float): current (mA/cm^2)
    """
    # voltage shift
    v = V + 10.0

    # steady state activation & inactivation
    mInf = 1.0 / (1.0 + e**(-(v + 1.0) / 12.0))
    hInf = 1.0 / (1.0 + e**(-(v + 54.0) / -11.0))

    # activation & inactivation time constants (ms)
    if v < -50:
        mTau = (1.25 + 175.03 * e**(0.026 * v)) / QT
    else:
        mTau = (1.25 + 13 * e**(-v * 0.026)) / QT

    hTau = (360 + (1010 + 24 * (v+55)) * e**(-((v + 75) / 48) ^ 2)) / QT

    m, h = state
    
    # cnexp update
    m = mInf + (m - mInf) * e**(-dt/mTau)
    h = hInf + (h - hInf) * e**(-dt/hTau)
    
    gK_PST = gK_Pstbar * m * m * h
    I_K = gK_PST * (V - E_K)
    return I_K

File: src/cpp/K_Pst.cpp
#include "K_Pst.hpp"

#include <cmath>

namespace {
const scalar_t e = 2.718281828459045235360287471352;
const scalar_t QT = std::pow(2.3, (34.0 - 21.0) / 10.0);
const scalar_t gK_Pstbar = 0.00001;
}

std::pair<scalar_t, scalar_t> K_Pst_derivatives(std::pair<scalar_t, scalar_t> state, scalar_t V) {
    const double v = V + 10.0;

    const scalar_t mInf = 1.0 / (1.0 + std::pow(e, (-(v + 1.0) / 12.0)));
    const scalar_t hInf = 1.0 / (1.0 + std::pow(e, (-(v + 54.0) / -11.0)));

    const scalar_t mTau = (v < -50.0) ? ((1.25 + 175.03 * std::pow(e, (0.026 * v))) / QT)
                                     : ((1.25 + 13.0 * std::pow(e, (-v * 0.026))) / QT);
    const scalar_t hTau = (360.0 + (1010.0 + 24.0 * (v + 55.0)) * std::pow(e, (-std::pow((v + 75.0) / 48.0, 2.0)))) / QT;

    const scalar_t m = state.first;
    const scalar_t h = state.second;

    const scalar_t dm_dt = (mInf - m) / mTau;
    const scalar_t dh_dt = (hInf - h) / hTau;

    return std::make_pair(dm_dt, dh_dt);
}

scalar_t K_Pst_current(std::pair<scalar_t, scalar_t> state, scalar_t V, scalar_t E_K, scalar_t dt) {
    const double v = V + 10.0;

    const scalar_t mInf = 1.0 / (1.0 + std::pow(e, (-(v + 1.0) / 12.0)));
    const scalar_t hInf = 1.0 / (1.0 + std::pow(e, (-(v + 54.0) / -11.0)));

    const scalar_t mTau = (v < -50.0) ? ((1.25 + 175.03 * std::pow(e, (0.026 * v))) / QT)
                                     : ((1.25 + 13.0 * std::pow(e, (-v * 0.026))) / QT);
    const scalar_t hTau = (360.0 + (1010.0 + 24.0 * (v + 55.0)) * std::pow(e, (-std::pow((v + 75.0) / 48.0, 2.0)))) / QT;

    scalar_t m = state.first;
    scalar_t h = state.second;

    m = mInf + (m - mInf) * std::pow(e, (-dt / mTau));
    h = hInf + (h - hInf) * std::pow(e, (-dt / hTau));

    const scalar_t gK_PST = gK_Pstbar * m * m * h;
    const scalar_t I_K = gK_PST * (V - E_K);
    return I_K;
}